### PR TITLE
Time ZK calls via ReportingZooKeeperClient

### PR DESCRIPTION
**Pull request by davidxia**
Make ReportingZooKeeperClient call ZooKeeperModelReport.time()
which gets a Timer from ZooKeeperMetrics and uses that to time
each ZK call.

### CODE REVIEW

**Comment by davidxia** [comment]
@spotify/helios-team

**Comment by davidxia** [comment]
Let me know if tests like this are worth having

```
@Test
  public void testEnsurePath() throws Exception {
    final ReportingZooKeeperClient reportingClient =
        new ReportingZooKeeperClient(client, reporter, TAG);
    reportingClient.ensurePath("path");
    verify(reporter).time(eq(TAG), eq("ensurePath"), Matchers.any(ZooKeeperCallable.class));
  }
```

File: helios-services/src/main/java/com/spotify/helios/servicescommon/coordination/ReportingZooKeeperClient.java [hunk 1]
```diff
@@ -51,122 +51,86 @@ public ReportingZooKeeperClient(final ZooKeeperClient client,
...
-      throw e;
-    }
+      return null;
+    });
   }
 
   @Override
   public void ensurePath(String path, boolean excludingLast) throws KeeperException {
-    try {
+    reporter.time(tag, "ensurePath", (ZooKeeperCallable<Void>) () -> {
```
Comment (mattnworb) [inline helios-services/src/main/java/com/spotify/helios/servicescommon/coordination/ReportingZooKeeperClient.java]: I don't think these casts to `(ZooKeeperCallable<Void>)` are necessary; or at least they seem to be in some spots in the diff and not others.

File: helios-services/src/main/java/com/spotify/helios/servicescommon/coordination/ReportingZooKeeperClient.java [hunk 2]
```diff
@@ -235,12 +177,7 @@ public Stat exists(String path) throws KeeperException {
 
   @Override
   public ZooKeeper.States getState() throws KeeperException {
-    try {
-      return client.getState();
-    } catch (KeeperException e) {
-      reporter.checkException(e, tag, "getState");
-      throw e;
-    }
+    return reporter.time(tag, "ensurePath", client::getState);
```
Comment (mattnworb) [inline helios-services/src/main/java/com/spotify/helios/servicescommon/coordination/ReportingZooKeeperClient.java]: should be "getState" here

File: helios-services/src/main/java/com/spotify/helios/servicescommon/coordination/ReportingZooKeeperClient.java (hunk 1 above)
Comment (davidxia) [inline helios-services/src/main/java/com/spotify/helios/servicescommon/coordination/ReportingZooKeeperClient.java]: You're right. IDE didn't catch these.

File: helios-services/src/main/java/com/spotify/helios/servicescommon/coordination/ReportingZooKeeperClient.java [hunk 3]
```diff
@@ -267,22 +204,14 @@ public CuratorFramework getCuratorFramework() {
...
 
   @Override
   public List<ACL> getAcl(final String path) throws KeeperException {
-    try {
-      return client.getAcl(path);
-    } catch (KeeperException e) {
-      reporter.checkException(e, tag, "getAcl");
-      throw e;
-    }
+    return reporter.time(tag, "ensurePath", () -> client.getAcl(path));
```
Comment (mattnworb) [inline helios-services/src/main/java/com/spotify/helios/servicescommon/coordination/ReportingZooKeeperClient.java]: "getAcl"

**Comment by mattnworb** [comment]
My own two cents is that it is not a very useful test of `ReportingZooKeeperClient` to verify that it called a certain method of the underlying client. It's more useful to test the values that those methods return.

What I would test though instead would be the new `Timer time(name, callable)` method in `ZooKeeperMetricsImpl` - you could have a test that asserts that a call like `.time("foo", ..)` adds a timer to the registry with the expected full name using the prefix.

**Comment by mattnworb** [comment]
also yay for java 8 language features :tada:

File: helios-services/src/main/java/com/spotify/helios/servicescommon/coordination/ZooKeeperCallable.java [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+
+import org.apache.zookeeper.KeeperException;
+
+/**
+ * A task that returns a result and may throw a {@link KeeperException}.
+ * We use this class when we don't want a {@link java.util.concurrent.Callable}
+ * that throws an Exception.
+ */
+@FunctionalInterface
+public interface ZooKeeperCallable<T> {
```
Comment (gimaker) [inline helios-services/src/main/java/com/spotify/helios/servicescommon/coordination/ZooKeeperCallable.java]: This is irrelevant for the function of the code but I would probably just have made this an inner class^Winterface of `ZooKeeperModelReporter` to reduce clutter - given how "big" this interface is :)

File: helios-services/src/main/java/com/spotify/helios/servicescommon/statistics/NoopZooKeeperMetrics.java [hunk 5]
```diff
@@ -17,7 +17,14 @@
...
 
+import com.codahale.metrics.Timer;
+
 public class NoopZooKeeperMetrics implements ZooKeeperMetrics {
   @Override
   public void zookeeperTransientError() {}
+
+  @Override
+  public Timer timer(String name) {
+    return new Timer();
```
Comment (gimaker) [inline helios-services/src/main/java/com/spotify/helios/servicescommon/statistics/NoopZooKeeperMetrics.java]: We shouldn't use a real timer, that actually gathers and keeps data, for a "no op" implementation. Unfortunately `Timer` isn't an interface so we can't really return a no-op timer...

Comment (gimaker) [inline helios-services/src/main/java/com/spotify/helios/servicescommon/statistics/NoopZooKeeperMetrics.java]: If you changed this interface slightly you could do an actual no-op implementation. E.g. something like:

```
void updateTimer(String name, final long durationMillis);
```

Comment (davidxia) [inline helios-services/src/main/java/com/spotify/helios/servicescommon/statistics/NoopZooKeeperMetrics.java]: It'd be nice to actually have a no op, but the nice thing about the current interface is that a Timer takes care of timing the duration. It's a bit more unwieldy if the caller has to handle timing a Callable when the Timer class already does this.

File: helios-services/src/main/java/com/spotify/helios/servicescommon/coordination/ZooKeeperCallable.java (hunk 4 above)
Comment (davidxia) [inline helios-services/src/main/java/com/spotify/helios/servicescommon/coordination/ZooKeeperCallable.java]: thanks

File: helios-services/src/main/java/com/spotify/helios/servicescommon/statistics/NoopZooKeeperMetrics.java (hunk 5 above)
Comment (davidxia) [inline helios-services/src/main/java/com/spotify/helios/servicescommon/statistics/NoopZooKeeperMetrics.java]: Hm, it might not be too bad. Let me try

**Pull request merged**